Added a naive orf research algorithm

althonos opened the pull request:
*Added an orf research algorithm.*

Matched orf are returned as **&[u8]**. The algorithm only searches for orf in the 3 possible reading frames of the provided sequence **without examining reverse complementary strand sequence**. If it did, then it would be necessary to calculate the reverse complementary strand, and orf couldn't be returned as slices because the reverse complementary strand sequence wouldn't live outside of the function.

The complexity of this algorithm is **linear**, the 3 reading frames are read simultaneously which means the sequence only needs to be parsed one time.

Refining the method could include **comparing the gc content** of the obtained orf to the gc content of the whole sequence (which is why a gc.rs is present in the seq_analysis folder).

johanneskoester commented on src/seq_analysis/gc.rs:
```diff
@@ -0,0 +1,43 @@
+const LOWER_G:u8 = 103;
+const LOWER_C:u8 = 99;
+
+const G:u8 = 71;
```
I believe you can write ``const G: u8 = b'G';``

johanneskoester commented on src/seq_analysis/gc.rs:
```diff
@@ -0,0 +1,43 @@
+        count += match sequence[i] {
+            C|LOWER_C|G|LOWER_G => 1f32, // G or C
+                              _ => 0f32,
+        } ;
```
Is the whitespace before the ; intentional?

johanneskoester commented on src/seq_analysis/gc.rs:
```diff
@@ -0,0 +1,43 @@
+const C:u8 = 67;
+
+
+pub fn gc <'a>(sequence: &'a [u8]) -> f32 {
```
Can you add a comment please?

johanneskoester commented on src/seq_analysis/gc.rs:
```diff
@@ -0,0 +1,43 @@
+        i += 3;
+    }
+    count / len as f32
+}
```
You can get rid of almost all type annotations here.

johanneskoester commented on src/seq_analysis/gc.rs:
```diff
@@ -0,0 +1,43 @@
+    count / len as f32
+}
+
+pub fn gc3 <'a>(sequence: &'a [u8]) -> f32 {
```
This counts the gc content of every third character? Why that? Also, it would be better to use [IntoIterator](https://doc.rust-lang.org/std/iter/trait.IntoIterator.html) as argument type. This would allow you to use the same function for both gc and gc3 (by modifying the iterator with a step_by).

johanneskoester commented:
Thanks! I wonder if there is are ORF finding algorithms described in the literature (pretty sure there are). If so, is your algorithm following one of these?

althonos commented:
Hi Johannes,

I'm not a bioinformatician, more a biologist on one hand and an
informatician on the other hand, and I'm quite new to Rust, so I'm not
totally aware of the language conventions.

I'm going to correct and add more comments  in the afternoon.

For the GC3 content, I read somewhere it was a quicker method than checking
for amino-acid usage in putative ORFs compared to already identified genes
but still more accurate than reading the whole GC content.

Concerning the algorithm, I'm no genius, that's the one I came with to only
read the full sequence one time (I took inspiration in the BioPython
library) but I'm sure someone already proposed it, I'll dig in the
literarure to find the reference.

I will try to look for more thorough algorithms to implement in the future
(like ones not taking start codons into account).

Regards, Martin.
Le 28 mars 2016 3:37 PM, "Johannes Köster" <notifications@github.com> a
écrit :

> Thanks! I wonder if there is are ORF finding algorithms described in the
> literature (pretty sure there are). If so, is your algorithm following one
> of these?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub
> <https://github.com/rust-bio/rust-bio/pull/48#issuecomment-202395911>
>

johanneskoester commented:
Dear Martin,  
sounds good. Let me know if you have questions regarding my comments so far. I'm positive regarding the current implementation. If possible I would like to have a reference to literature, similar to many other algorithms available in Rust-Bio.

Johannes

althonos commented:
Dear Johannes,

I added some comments and corrected the code for the sake of clarity.

I also factored the code for the gc and gc3 counter.
However, I chose to stick to the slice type for input sequence, as some 
searches on the Internet showed that the step_by method wasn't currently 
optimized, so a plain old while loop stays in my opinion the best 
solution in term of performances (and because the type of items 
contained by an IntoIterator type couldn't be defined at compile time, 
the compiler refused to perform any comparison later in the code).

I couldn't find any literature reference to the algorithm however, as 
all the papers I read reference an /idea /instead of an /implementation 
/(or it might also be that I don't know where to look, I tried PubMed & 
Google Scholar).

I may be adding another algorithm to look for Maximum Length Open 
Reading Frame (slices of sequence not starting by a methionine but only 
considering the space between two STOP codons). Do you want me to 
implement it before any merge ? It should be done by tomorrow.

And I forgot to tell you, but as I'm not working with rust nightly 
compiler, I couldn't check the benches I wrote, so I'm not sure if they 
compile and/or work.

Regards, Martin.

On 28/03/2016 17:52, Johannes Köster wrote:
>
> Dear Martin,
>
> sounds good. Let me know if you have questions regarding my comments 
> so far. I'm positive regarding the current implementation. If possible 
> I would like to have a reference to literature, similar to many other 
> algorithms available in Rust-Bio.
>
> Johannes
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub 
> <https://github.com/rust-bio/rust-bio/pull/48#issuecomment-202452548>
>

johanneskoester commented on README.md:
```diff
@@ -21,6 +21,7 @@ Currently, rust-bio provides
 * BWT and FM-Index,
 * FMD-Index for finding supermaximal exact matches,
 * a q-gram index,
+* an orf research algorithm
```
Do you mean "ORF search algorithm"?

althonos replied:
Yes, my bad =="
Le 29 mars 2016 15:31, "Johannes Köster" <notifications@github.com> a
écrit :

> In README.md
> <https://github.com/rust-bio/rust-bio/pull/48#discussion_r57722639>:
>
> > @@ -21,6 +21,7 @@ Currently, rust-bio provides
> >  * BWT and FM-Index,
> >  * FMD-Index for finding supermaximal exact matches,
> >  * a q-gram index,
> > +* an orf research algorithm
>
> Do you mean "ORF search algorithm"?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub
> <https://github.com/rust-bio/rust-bio/pull/48/files/5ccc54036a28e80c7729883d00d2431991da01c7#r57722639>
>

johanneskoester commented on src/seq_analysis/gc.rs:
```diff
@@ -0,0 +1,33 @@
+// Maybe not the most clever way to do it, but at least runs in o(n)
+
+/// Base gc content counter
+fn gcn_content<'a>(sequence: &'a[u8], step: usize) -> f32 {
```
I have added trait aliases that allow to easily use IntoIterator (see [here](http://rust-bio.github.io/rust-bio/bio/utils/index.html)). This way, you can get rid of the while loop. Using ``step`` from Itertools should be fine, and I don't think it will hurt performance.

johanneskoester commented on src/seq_analysis/orf.rs:
```diff
@@ -0,0 +1,152 @@
+        }
+    }
+    /// Find all open reading frames in given sequence. Matches are returned as iterators over vectors
+    pub fn find_all(&'a self, sequence: &'a [u8]) -> Matches {
```
Can you use the new ``TextSlice`` type alias here (see bio::utils)? Of course, since your algorithm is O(n) it would be even better to use ``IntoTextIterator``, but only if the code does not become over complicated.

johanneskoester commented on src/seq_analysis/orf.rs:
```diff
@@ -0,0 +1,152 @@
+                [ self.sequence[self.state.index]  , self.sequence[self.state.index+1], self.sequence[self.state.index+2] ],
+                [ self.sequence[self.state.index+1], self.sequence[self.state.index+2], self.sequence[self.state.index+3] ],
+                [ self.sequence[self.state.index+2], self.sequence[self.state.index+3], self.sequence[self.state.index+4] ],
+            ];
```
Above is not very efficient and repetetive. You can use slicing instead, i.e. ``self.sequence[self.state.index..self.state.index+3]``. To avoid the repetition, create a private method that takes an offset and returns ``self.sequence[offset..offset + 3]``.

johanneskoester replied:
This is even better: You maintain a [deque](https://doc.rust-lang.org/collections/vec_deque/struct.VecDeque.html) of the last 5 nucleotides in the struct. In each step, you ``self.codons.pop_front()`` the first nucleotide, and ``self.codons.push_back()`` the new nucleotide. It should also allow you to remove the dependence on having a slice of text, and allow you to use a ``IntoTextIterator`` instead.

althonos replied:
Working on it, thanks for the hints :)
Le 29 mars 2016 20:34, "Johannes Köster" <notifications@github.com> a
écrit :

> In src/seq_analysis/orf.rs
> <https://github.com/rust-bio/rust-bio/pull/48#discussion_r57775970>:
>
> > +
> > +impl<'a> Iterator for Matches<'a> {
> > +    type Item = (usize, usize, &'a [u8]);
> > +
> > +    fn next(&mut self) -> Option<(usize, usize, &'a [u8])> {
> > +
> > +        let mut orf: Option<(usize, usize, &'a [u8])> = None;
> > +        let mut codons: [ [u8; 3]; 3]; //
> > +
> > +        while self.state.index < self.state.length - 3 {
> > +            //codons array depending on reading frame
> > +            codons = [
> > +                [ self.sequence[self.state.index]  , self.sequence[self.state.index+1], self.sequence[self.state.index+2] ],
> > +                [ self.sequence[self.state.index+1], self.sequence[self.state.index+2], self.sequence[self.state.index+3] ],
> > +                [ self.sequence[self.state.index+2], self.sequence[self.state.index+3], self.sequence[self.state.index+4] ],
> > +            ];
>
> This is even better: You maintain a deque
> <https://doc.rust-lang.org/collections/vec_deque/struct.VecDeque.html> of
> the last 5 nucleotides in the struct. In each step, you
> self.codons.pop_front() the first nucleotide, and self.codons.push_back()
> the new nucleotide. It should also allow you to remove the dependence on
> having a slice of text, and allow you to use a IntoTextIterator instead.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub
> <https://github.com/rust-bio/rust-bio/pull/48/files/5ccc54036a28e80c7729883d00d2431991da01c7#r57775970>
>

johanneskoester commented:
Hi! Any updates?

althonos commented:
Hi !
I worked on the algorithm so now it should be able to accept any iterator,
now i'm working on getting the inside of the loop to work the right way...
I didn't commit though, as it doesn't work right now :)
Le 20 avr. 2016 11:23, "Johannes Köster" <notifications@github.com> a
écrit :

Hi! Any updates?

—
You are receiving this because you authored the thread.
Reply to this email directly or view it on GitHub
<https://github.com/rust-bio/rust-bio/pull/48#issuecomment-212345103>

johanneskoester commented:
Sure. Let me know whenever you are ready.

althonos commented:
Hi Johannes,

well I have an issue, I was trying to port the code to use the 
IntoTextIterator trait you aliased, but the structure is becoming quite 
complex because :
     - the Match struct has the Iterator trait, returned Items are Orf
       (so that in "for n in finder.find_all(sequence) {}", n takes 
value of every Orf the Finder could find).

But as you know, the IteratorTrait requires the implementation of "fn 
next(&mut self)", which means it is impossible to consume another 
iterator in next().

I thought about implementing the IntoIterator trait for the Match 
struct, and then have Match.into_iter() return a custom Iterator which 
would itself contain an iterator over the sequence, but I have several 
issues with Traits concerning the IntoIterator implementation for the 
Match struct.

So I don't know if it is such a good idea to return the Orf as items of 
an Iterator, because if we did collect all Orf into a Vec<Orf>, it would 
be simpler to implement a method in the Finder struct.

Best Regards,
Martin.


On 20/04/2016 11:23, Johannes Köster wrote:
>
> Hi! Any updates?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub 
> <https://github.com/rust-bio/rust-bio/pull/48#issuecomment-212345103>
>

johanneskoester commented:
Hi,
thanks for trying to adapt to these latest changes! I think looking at [this example](https://github.com/rust-bio/rust-bio/blob/master/src/pattern_matching/myers.rs#L103) might clarify how to use `IntoTextIterator`. Let me know if I did not get the problem or in case you need further explanation.

Thanks again! I am glad to have you contributing!

althonos commented:
Thanks for the example, I shamelessly copied and adapted it so now my inner loop works :D
It's just a matter of time before I complete the algorithm now !

althonos commented:
Hi, the implementation is complete on my side, I'm awaiting your review ! :smile: 
Don't know why Travis fails though...

zitsen commented on src/seq_analysis/orf.rs:
```diff
@@ -0,0 +1,161 @@
+//! let min_len:usize = 50;
+//! let finder = Finder::new(start_codons, stop_codons, min_len);
+//!
+//! let sequence = b"ACGGCTAGAAAAGGCTAGAAAA"
```
Lost a `;` at the end of line 19.

zitsen commented:
IMO, as far as I know, use a `rebase` from upstream master instead of `merge` would be much better, though I don't know if rust-bio developers care about this.

althonos commented:
I'll change that and run a rebase, although the current tree is already a
mess x)
Thanks !
Le 19 mai 2016 3:48 AM, "Huo Linhe" <notifications@github.com> a écrit :

> IMO, as far as I know, use a rebase from upstream master instead of merge
> would be much better, though I don't know if rust-bio developers care about
> this.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub
> <https://github.com/rust-bio/rust-bio/pull/48#issuecomment-220207101>
>

althonos commented:
Done.
Le 19 mai 2016 03:48, "Huo Linhe" <notifications@github.com> a écrit :

> IMO, as far as I know, use a rebase from upstream master instead of merge
> would be much better, though I don't know if rust-bio developers care about
> this.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub
> <https://github.com/rust-bio/rust-bio/pull/48#issuecomment-220207101>
>

zitsen commented on src/seq_analysis/orf.rs:
```diff
@@ -0,0 +1,161 @@
+//! let sequence = b"ACGGCTAGAAAAGGCTAGAAAA";
+//!
+//! for orf in finder.find_all(sequence) {
+//!    ...do something...
```
comment this line as `  // ... do something ...` to pass test

zitsen commented on src/seq_analysis/orf.rs:
````diff
@@ -0,0 +1,161 @@
+//! # Example
+//!
+//! ```
+//! use bio::seq_analysis::orf::NaiveFinder;
````
Is this `Finder` not `NaiveFinder`?

johanneskoester commented on src/seq_analysis/orf.rs:
```diff
@@ -0,0 +1,161 @@
+                    self.state.orf[offset].extend(self.state.codon.clone().into_iter());
+                    // build results
+                    if self.state.orf[offset].len() > self.finder.min_len {
+                        result = Some((offset,
```
Why can't we just return an interval with start and stop (exclusive) in the given sequence. This way, we don't need the allocation for the orf sequence. If people want the orf sequence, they can slice the original sequence with the returned interval.

johanneskoester commented:
Personally I am not a fan of rebase. So, in the future, you can omit that.

johanneskoester commented:
The code looks quite good. I have added one comment above.

althonos commented:
IMO the allocation of the newfound sequences is better. Because the current
implementation uses an IntoIterator, that means the full sequence will be
consumed when you call the function (finder.find_all takes ownership and
then consumes the iterator). So it would be needed to slice the orf from
the original sequence to have a backup, which means either reparsing the
file/sequence or running the orf search on a cloned Text, both meaning
double allocation of the whole sequence. So instead of having to reallocate
the full sequence here whe only allocate the ones that matter.

Still, I get your point, so maybe it would be possible to wrap the orf
sequence into an Option<Text>, and then have a boolean as an argument of
Finder::new() to set it the allocation of the orf is wanted. But that may
also raise the complexity of the items in the iterator.

The code looks quite good. I have added one comment above.

—
You are receiving this because you authored the thread.
Reply to this email directly or view it on GitHub
<https://github.com/rust-bio/rust-bio/pull/48#issuecomment-220246228>

johanneskoester commented:
Actually, IntoTextIterator and the corresponding TextIterator where explicitly designed to not consume the given sequence. That would be totally not desirable. They already iterator over references of `u8`. Hence, you will always be able to just give a reference to the sequence to the orf finder.

althonos commented:
Ok then, I'll remove the allocation in the function's loop, I'll tell you
when it's done (tonight eventually).
Le 19 mai 2016 11:19 AM, "Johannes Köster" <notifications@github.com> a
écrit :

> Actually, IntoTextIterator and the corresponding TextIterator where
> explicitly designed to not consume the given sequence. That would be
> totally not desirable. They already iterator over references of u8.
> Hence, you will always be able to just give a reference to the sequence to
> the orf finder.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub
> <https://github.com/rust-bio/rust-bio/pull/48#issuecomment-220271452>
>

althonos commented:
Done.
I ran a bench based on allocation of found sequences, with in algorithm
allocation or slicing from total sequence.

In algorithm allocation ran in around 700,000ns/iter, slicing ran in around
390,000 ns/iter (see updated bench_orf).

I wonder if it's still useful to include the offset of the orf in the
returned tuple, however if one day we are to write a Finder performing
simultaneously reverse complementation and orf search, maybe it would be
best to leave it for consistency (so that returned tuples would be (-1,x,y)
or (+3,x,y)).

Ok then, I'll remove the allocation in the function's loop, I'll tell you
when it's done (tonight eventually).
Le 19 mai 2016 11:19 AM, "Johannes Köster" <notifications@github.com> a
écrit :

> Actually, IntoTextIterator and the corresponding TextIterator where
> explicitly designed to not consume the given sequence. That would be
> totally not desirable. They already iterator over references of u8.
> Hence, you will always be able to just give a reference to the sequence to
> the orf finder.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub
> <https://github.com/rust-bio/rust-bio/pull/48#issuecomment-220271452>
>

johanneskoester commented:
Sure, you should leave the offset. I think if it would make sense to return a struct though (since tuples become error prone with three values):

```
pub struct ORF {
    pub start: usize,
    pub end: usize,
    pub offset: u8
}
```

What values can offset have?

althonos commented:
Yes, I thought about creating a struct too. Offset is either +1,+2 or +3,
in the current situation, and is intended to be -1, -2 or -3 if the orf is
found on the reverse strand.
Le 19 mai 2016 5:19 PM, "Johannes Köster" <notifications@github.com> a
écrit :

> Sure, you should leave the offset. I think if it would make sense to
> return a struct though (since tuples become error prone with three values):
>
> pub struct ORF {
>     pub start: usize,
>     pub end: usize,
>     pub offset: u8
> }
>
> What values can offset have?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub
> <https://github.com/rust-bio/rust-bio/pull/48#issuecomment-220357525>
>

johanneskoester commented:
Ok, then the struct makes sense as I proposed it I guess. Feel free to
implement it that way.  
  

  

  

On May 19 2016, at 5:39 pm, Martin Larralde &lt;notifications@github.com&gt;
wrote:  

> Yes, I thought about creating a struct too. Offset is either +1,+2 or +3,  
in the current situation, and is intended to be -1, -2 or -3 if the orf is  
found on the reverse strand.  
Le 19 mai 2016 5:19 PM, "Johannes Köster" &lt;notifications@github.com&gt; a  
écrit :  
  
&gt; Sure, you should leave the offset. I think if it would make sense to  
&gt; return a struct though (since tuples become error prone with three
values):  
&gt;  
&gt; pub struct ORF {  
&gt; pub start: usize,  
&gt; pub end: usize,  
&gt; pub offset: u8  
&gt; }  
&gt;  
&gt; What values can offset have?  
&gt;  
&gt; —  
&gt; You are receiving this because you authored the thread.  
&gt; Reply to this email directly or view it on GitHub  
&gt; &lt;https://github.com/rust-bio/rust-
bio/pull/48#issuecomment-220357525&gt;  
&gt;  

>

> —  
You are receiving this because you commented.  
Reply to this email directly or [view it on GitHub](https://github.com/rust-
bio/rust-bio/pull/48#issuecomment-220363969)![](https://github.com/notificatio
ns/beacon/ABxcVti_wCpD9lPVQHhBUM99nfQymasKks5qDIQGgaJpZM4H5fVx.gif)

althonos commented:
Except offset has got to be an i8 instead of an u8 :)
Le 19 mai 2016 10:36 PM, "Johannes Köster" <notifications@github.com> a
écrit :

> Ok, then the struct makes sense as I proposed it I guess. Feel free to
> implement it that way.
>
>
>
>
>
>
> On May 19 2016, at 5:39 pm, Martin Larralde &lt;notifications@github.com
> &gt;
> wrote:
>
> > Yes, I thought about creating a struct too. Offset is either +1,+2 or
> +3,
> in the current situation, and is intended to be -1, -2 or -3 if the orf is
> found on the reverse strand.
> Le 19 mai 2016 5:19 PM, "Johannes Köster" &lt;notifications@github.com&gt;
> a
> écrit :
>
> &gt; Sure, you should leave the offset. I think if it would make sense to
> &gt; return a struct though (since tuples become error prone with three
> values):
> &gt;
> &gt; pub struct ORF {
> &gt; pub start: usize,
> &gt; pub end: usize,
> &gt; pub offset: u8
> &gt; }
> &gt;
> &gt; What values can offset have?
> &gt;
> &gt; —
> &gt; You are receiving this because you authored the thread.
> &gt; Reply to this email directly or view it on GitHub
> &gt; &lt;https://github.com/rust-bio/rust-
> bio/pull/48#issuecomment-220357525&gt;
> &gt;
>
> >
>
> > —
> You are receiving this because you commented.
> Reply to this email directly or [view it on GitHub](
> https://github.com/rust-
> bio/rust-bio/pull/48#issuecomment-220363969)![](
> https://github.com/notificatio
> ns/beacon/ABxcVti_wCpD9lPVQHhBUM99nfQymasKks5qDIQGgaJpZM4H5fVx.gif)
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub
> <https://github.com/rust-bio/rust-bio/pull/48#issuecomment-220440833>
>

althonos commented:
Okay, now Orfs are returned as a struct, I updated the bench and the test in orf.rs.

johanneskoester commented on src/seq_analysis/orf.rs:
```diff
@@ -0,0 +1,163 @@
+    pub fn new() -> Self {
+        State {
+            start_pos: [0, 0, 0],
+            in_orf: [false, false, false],
```
One final issue, sorry! I think it would be more rustic to write ``start_pos: [None, None, None]``. Then, you can insert start positions with ``Some(0)``, and replace the if else below with a match construct.

althonos replied:
Done !
Le 21 mai 2016 7:36 PM, "Johannes Köster" <notifications@github.com> a
écrit :

> In src/seq_analysis/orf.rs
> <https://github.com/rust-bio/rust-bio/pull/48#discussion_r64137711>:
>
> > +}
> > +
> > +/// The current algorithm state.
> > +struct State {
> > +    start_pos: [usize; 3],
> > +    in_orf: [bool; 3],
> > +    codon: VecDeque<u8>,
> > +}
> > +
> > +
> > +impl State {
> > +    /// Create new state.
> > +    pub fn new() -> Self {
> > +        State {
> > +            start_pos: [0, 0, 0],
> > +            in_orf: [false, false, false],
>
> One final issue, sorry! I think it would be more rustic to write start_pos:
> [None, None, None]. Then, you can insert start positions with Some(0),
> and replace the if else below with a match construct.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub
> <https://github.com/rust-bio/rust-bio/pull/48/files/2e4800f3ffe5ed6f0a6b89d951ef691ea6c842c6#r64137711>
>

johanneskoester commented on src/seq_analysis/orf.rs:
```diff
@@ -0,0 +1,163 @@
+            offset = (index + 1) % 3;
+
+            // inside orf
+            if self.state.in_orf[offset] {
```
Since you have now options in start_pos, in_orf is not needed anymore. You can instead check if the option is none.

althonos commented:
Okay, removed in_orf attribute.

johanneskoester commented:
Awesome, thanks! I think it is time to merge!
